compiler/damlc/daml-stdlib-src/DA/Internal/Template.daml

```diff
@@ -72,91 +54,45 @@ fetchByKey : forall c k. TemplateKey c k => k -> Update (ContractId c, c)
+-- The `t` needs to be passed using an explicit type application. For
 -- instance, if you want to exercise a choice `Withdraw` on a contract of
 -- template `Account` given by its key `k`, you must call
 -- `exerciseByKey @Account k Withdraw`.
-exerciseByKey : forall c k e r. (TemplateKey c k, Choice c e r) => k -> e -> Update r
-exerciseByKey k e = do
-    (cid, _) <- fetchByKey @c k
-    internalExercise cid e
-
-class Template c where
-
-    -- | Predicate that must hold for the succesful creation of the contract.
-    ensure : c -> Bool
-    ensure _ = True
-
-    -- | The signatories of a contract.
-    signatory : c -> [Party]
-
-    -- | The observers of a contract.
-    observer : c -> [Party]
-    observer _ = []
-
-    -- | The agreement text of a contract.
-    agreement : c -> Text
-    agreement _ = ""
-
-    -- | HIDE
-    internalCreate : c -> Update (ContractId c)
-    internalCreate = magic @"create"
-
-    -- | HIDE
-    internalFetch : ContractId c -> Update c
-    internalFetch = magic @"fetch"
+exerciseByKey : forall t k c r. (TemplateKey t k, Choice t c r) => k -> c -> Update r
+exerciseByKey k c = do
+    (cid, _) <- fetchByKey @t k
+    exercise cid c
 
-    -- | HIDE
-    internalArchive : ContractId c -> Update ()
-    internalArchive = magic @"archive"
+data Archive = Archive{}
+  deriving (Eq, Show)
 
+data NonConsuming t = NonConsuming {}
+data PreConsuming t = PreConsuming {}
+data PostConsuming t = PostConsuming {}
 
--- Deliberately not exported.
-data Consuming = PreConsuming  -- Archive before executing exercise body.
-               | PostConsuming -- Execute exercise body then archive.
-               | NonConsuming  -- Don't archive.
+data HasKey t = HasKey {}
 
--- | HIDE
-nonconsuming : NoEvent c e -> Consuming
-nonconsuming _ = NonConsuming
-
--- | HIDE
-preconsuming : NoEvent c e -> Consuming
-preconsuming _ = PreConsuming
-
--- | HIDE
-postconsuming : NoEvent c e -> Consuming
-postconsuming _ = PostConsuming
-
-class Template c => Choice c e r | c e -> r where
-    -- | HIDE
-    consuming : NoEvent c e -> Consuming
-    consuming = preconsuming
-
-    choiceController : c -> e -> [Party]
-    choice : c -> ContractId c -> e -> Update r
+-- FIXME(RJR): Do we need these?
+-- nonconsuming : NoEvent c e -> NonConsuming
+-- nonconsuming _ = NonConsuming
 
-    -- | HIDE
-    internalExercise : ContractId c -> e -> Update r
-    internalExercise = magic @"exercise"
+-- preconsuming : NoEvent c e -> PreConsuming
+-- preconsuming _ = PreConsuming
 
-
--- | The data type corresponding to the implicit `Archive`
--- choice in every template.
-data Archive = Archive
-
-instance Template c => Choice c Archive () where
-    choiceController c _ = signatory c
-    choice _ _ _ = return ()
-
-    internalExercise c Archive = internalArchive c
+-- postconsuming : NoEvent c e -> PostConsuming
+-- postconsuming _ = PostConsuming
 
 -- | HIDE
-data NoEvent c e = NoEvent
+-- data NoEvent c e = NoEvent
```

Comment:
Let's remove it if we don't need it anymore.

Reply:
Done